### PR TITLE
Refresh token

### PR DESCRIPTION
**Pull request by ciremusyoka**
Part of https://github.com/onaio/reveal-frontend/issues/1439

**Changes included with this PR**
- Add expected time for access and refresh token expiry
- Add selectors for getting access and refresh token status
- Add a default component to be displayed when token is expired
- Redirect user to above page if token is expired 

**Checklist**

- [x] tests are included and passing
- [x] documentation is changed or added
- [ ] a .storybook story was added where necessary.
- [x] code is [transpiled](https://github.com/onaio/js-tools#transpiling-the-package-and-generating-type-definations-for-it)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -119,6 +119,7 @@
     "webpack": "^4.29.6"
   },
   "dependencies": {
+    "mockdate": "^3.0.2",
```
Comment (moshthepitt) [inline package.json:122]: Should this be moved to dev dependencies?

File: packages/gatekeeper/package.json [hunk 2]
```diff
@@ -31,14 +31,16 @@
   "author": "Ona Engineering",
   "license": "Apache-2.0",
   "dependencies": {
-    "@onaio/session-reducer": "^0.0.11",
+    "@onaio/session-reducer": "../js-tools/packages/session-reducer/",
```
Comment (moshthepitt) [inline packages/gatekeeper/package.json:34]: Why?

File: packages/gatekeeper/src/components/TokenExpiry/constants.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export const LOGIN_BUTTON_TEXT = 'Renew session';
+export const SESSION_EXPIRED_TEXT =
+  'Your session has expired. Please click the link below to start a new session';
```
Comment (moshthepitt) [inline packages/gatekeeper/src/components/TokenExpiry/constants.ts:3]: Please change to something like "Your session has expired. Please click the link below to log in again."

File: packages/gatekeeper/src/helpers/oauth.ts [hunk 4]
```diff
@@ -49,16 +58,41 @@ export function getOnadataUserInfo(apiResponse: { [key: string]: any }): Session
   };
 }
 
+/**
+ * add given second to current time
```
Comment (moshthepitt) [inline packages/gatekeeper/src/helpers/oauth.ts:62]: Could you describe the purpose and usage of this better in the docstring here?

File: packages/session-reducer/src/index.ts [hunk 5]
```diff
@@ -165,6 +210,25 @@ export function getAccessToken(state: Partial<Store>): string | null {
...
+ */
+export function isTokenExpired(state: Partial<Store>): boolean {
+  return !getAccessToken(state) ? true : getTokenExiryStatus(state) === TokenStatus.expired;
+}
+
+/**
+ * get token or redirect to session expired page
+ * @param {Partial<Store>} state - the redux store
+ */
+export const getTokenOrRedirect = (state: Partial<Store>, redirectTo: string | null = null) => {
```
Comment (moshthepitt) [inline packages/session-reducer/src/index.ts:224]: Do you think that the session reducer package needs to be the entity that determines what happens when a session is not valid?  Why?

Please try to change this opinion that I currently hold: the session reducer package does not care what you do, it just gives you the token, or tells you that the current token is invalid.

File: packages/gatekeeper/package.json (hunk 2 above)
Comment (ciremusyoka) [reply packages/gatekeeper/package.json]: was a mistake

File: packages/session-reducer/src/index.ts [hunk 6]
```diff
@@ -146,6 +153,43 @@ export function getUser(state: Partial<Store>): User {
...
+export function getRefreshTokenExpiryStatus(state: Partial<Store>): TokenStatus {
+  const extraData = (state as any)[reducerName].extraData;
+  if (extraData.oAuth2Data && extraData.oAuth2Data.refresh_expires_at) {
+    return new Date(Date.now()) >= new Date(extraData.oAuth2Data.refresh_expires_at)
+      ? TokenStatus.expired
+      : TokenStatus.active;
+  }
+  return TokenStatus.timeNotFound;
+}
+
```
Comment (moshthepitt) [inline packages/session-reducer/src/index.ts:192]: Could we make this more DRY?

File: packages/gatekeeper/src/components/TokenExpiry/index.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+/** TokenExpired props interface */
+export interface TokenExpiredProps {
+  logoutLinkText: string; // text on the logout link button
+  logoutUrl: string;
+  sessionExpiryText: string; // text informing users their session has expired
+}
+
+/** default page to display when token expires */
+const TokenExpired = (props: TokenExpiredProps) => {
```
Comment (moshthepitt) [inline packages/gatekeeper/src/components/TokenExpiry/index.tsx:14]: Where is this component used?

Comment (ciremusyoka) [reply packages/gatekeeper/src/components/TokenExpiry/index.tsx:14]: will act as a default component which user can be redirected to when a token is expired.

Comment (moshthepitt) [reply packages/gatekeeper/src/components/TokenExpiry/index.tsx:14]: Should we add that to the docs of gatekeeper?

File: jest.config.js [hunk 8]
```diff
@@ -1,3 +1,5 @@
+process.env.TZ = 'GMT';
```
Comment (moshthepitt) [inline jest.config.js:1]: Can we use our own timezone maybe? :)

**Review by moshthepitt — APPROVED** [review]

**Review by KipSigei — APPROVED** [review]

**Pull request merged**